utils/config.go

```diff
@@ -144,7 +144,7 @@ func GetLogFileLocation(fileLocation string) string {
 }
 
 func SaveConfig(fileName string, config *model.Config) *model.AppError {
-	b, err := json.MarshalIndent(config, "", "    ")
+	b, err := json.MarshalIndent(config.SaveClone(), "", "    ")
```

Comment:
It took me a while before I realized what `SaveClone` actually did. I'm worried about the presence of such a function. Is there a reason we don't just add `json:"-"` to the `SupportedTimezones` field on the config to prevent its serialization? Or are we serializing the full configuration in other contexts?

Reply:
I've tried 2 other approaches

1. json:"-"
This avoids Deserialization and Serialization. Therefore, pulling `timezones.json` wouldn't work.

2. SupportedTimezones implementing the json.Marshal interface.

If I'm recalling this perfectly:
```
func (SupportedTimezones) MarshalJSON() ([]byte, err) {
   return []byte("[]",) nil
}
```

Works but you end up with this in `config.json`
```
"SupportedTimezones": []
```

Not pretty 😖 

Therefore, creating a new struct that serializes only what we need was the best option.  Actually the preferred solution in the community.  It get's the job done, and avoids saving extraneous data

Reply:
Ah, I understand now: `LoadTimezoneConfig` is deserializing into a `Config` data structure, vs. serializing into a structure of it's own. 

I agree that creating a new struct for custom serialization is often the right idea, but in this case it seems forced. `SupportedTimezones` is made to be part of the config structure but then excluded in almost every way. It has its own `LoadTimezoneConfig` that doesn't actually load the `TimezoneSettings` part of the `Config`, just the `SupportedTimezones`.

Have we considered making `SupportedTimezones` not part of the `Config`, but a sibling on the `App`? I'm 1/5 on that approach.

Or, what if `LoadTimezoneConfig` was written to just return a `SupportedTimezones` struct instead of a `Config` object, allowing `LoadConfig` to basically work the same way, but then trivially excluding the `SupportedTimezones` struct with the `json:"-"` when serializing to the client? Seems like the best of all worlds.

Apart from the modelling concerns above, I think my concern about the `SaveClone` method stems from the fact that:
1) It's not obvious that this function only excludes a single struct from the Config (you have to scan to ask, "What's missing?")
2) It seems easy to forget to update this method when adding a new configuration field that is meant to be saved.

My apologies for not providing this feedback earlier -- I hadn't been following the conversation on this pull request closely until @jwilander kindly poked me :)

Reply:
No problem, it can be hard to understand the issue without context.

Could you bear with me for a second, I quickly gave your approach a shot.

1. `SupportedTimezones    SupportedTimezones json:"-"` in `model/config.go  type Config`

2. Here's the `LoadTimezones` function
```
func LoadTimezones(fileName string) (timezones model.SupportedTimezones, configPath string, appErr *model.AppError) {
	config, _, _ := loadConfigFile(fileName)
	return config.SupportedTimezones, "", nil
}
```

3. And in `app/config.go` the transfer process
```
supportedTimezones, _, err := utils.LoadTimezones(timezonePath)
if err != nil {
	return err
}

cfg.SupportedTimezones = supportedTimezones
```

This is where it gets rather odd.
When I call the `system/timezones` endpoint.

I get no timezones.  The array is empty.
If I remove `json:"="` it works.

My assumptions about `json` are wrong.  I thought it only affects serialize processing.  Am I wrong? I may be missing something

Reply:
Adding the `json:'-'` struct tag to the Config struct will affect any serialization and deserialization of that, but I don't think it should affect serializing the value of the field. 

Is this what you were thinking, Jesse? https://gist.github.com/hmhealey/ff79c43a3fce0dc1d1a87e1a35d4f399

Reply:
@hmhealey, yeah that's the spirit of what I had hoped would simplify the code for @miguelespinoza.

The reason `timezones` is empty above is because `LoadTimezones` in your example still depends on deserializing the Config structure, in which the `SupportedTimezones` field is now marked as ignored. If we agree this is the right direction, I'm hoping we can decouple the supported timezones from config as much as possible, landing on something less tied to the existing config parsing, ala:

```go
func LoadTimezones(fileName string) (model.SupportedTimezones, *model.AppError) {
          var supportedTimezones model.SupportedTimezones

          if timezoneFile := FindConfigFile(fileName); timezoneFile == "" {
                  return nil, model.NewAppError("LoadTimezones", "no.such.file", map[string]interface{}{"Filename": fileName}, "", 0)
          } else if raw, err := ioutil.ReadFile(timezoneFile); err != nil {
                  return nil, model.NewAppError("LoadTimezones", "failed.to.open.file", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
          } else if err := json.Unmarshal(raw, &supportedTimezones); err != nil {
                  return nil, model.NewAppError("LoadTimezones", "not.json", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
          } else {
                  return supportedTimezones, nil

          }
  }
```

Of course, this also implies flattening `timezones.json`:
```
[
    "Africa/Abidjan",
    "Africa/Accra",
    "Africa/Addis_Ababa",
    "Africa/Algiers",
    "Africa/Asmara",
    "Africa/Asmera",
    "Africa/Bamako",
    "Africa/Bangui",
    "Africa/Banjul",
...
]
```

Reply:
@hmhealey and @lieut-data, I see, thanks. I can take the approach of decoupling the timezone from the config.

Just for some guidance would this live in `app.go`, following the same pattern as config or license?
Basically, `timezone       atomic.Value`
Or is there a simpler path for supporting timezones.  The goal is to call `systems/timezones` which was currently in this format

```
func getSupportedTimezones(c *Context, w http.ResponseWriter, r *http.Request) {
	supportedTimezones := c.App.Config().SupportedTimezones

	if len(supportedTimezones) == 0 {
		emptyTimezones := make([]string, 0)
		w.Write([]byte(model.TimezonesToJson(emptyTimezones)))
		return
	}

	w.Write([]byte(model.TimezonesToJson(supportedTimezones)))
}
```

I imagine decoupling the timezone would look something like
`supportedTimezones := c.App.Timezones()`

Reply:
This would be the last thing, hopefully, once we can come to an agreement I'll get this in and we can get this ready for 4.9 😄

Reply:
That sounds like it would work well. I think the problems we're running into here are mostly caused by having the timezones as a kind-of-but-not-really part of the config. Separating them out should fix that